### PR TITLE
Clean up in unet.py

### PR DESCRIPTION
**Pull request by bil-y**
- address left-overs from previous unet MR
- address #101 
- address #100 
- more to come

### CODE REVIEW

**Comment by phdenzel** [comment]
> @mstadelmann: as discussed offline, there is something wrong with `num_layers_per_block>1`. Perhaps you have already found a solution, otherwise, I'll work on it tomorrow.

Sorry, forgot to add the actual changes... now the tests actually fail 😅

**Comment by phdenzel** [comment]
I think, it's as simple as commenting out lines 280-281:

```python
if skip_connection_to_all_layers is None:
     skip_connection_to_all_layers = skip_connection_action == "concat"
```

File: src/chuchichaestli/models/unet/unet.py [hunk 1]
```diff
@@ -224,19 +265,23 @@ def __init__(
...
+        if skip_connection_to_all_layers is None:
+            skip_connection_to_all_layers = skip_connection_action == "concat"
 
-        if skip_connection_between_levels is None:
-            skip_connection_between_levels = skip_connection_action == "concat"
+        # Build decoder
+        self.up_blocks = nn.ModuleList([])
 
         for i in reversed(range(n_mults)):
             ins = outs
```
[the remark this answers is not in the document]
Comment (mstadelmann) [reply src/chuchichaestli/models/unet/unet.py:287]: And basically this raises the question of how to deal with the skip connections if we have multiple blocks again.

If we have skip_connection_action = concat
- Do we concatenate the out of the last corresponding down block to all up blocks?

I guess the idea was to control this with skip_connection_to_all_levels?

Maybe we want to rename "skip_connection_to_all_levels" to "skip_connection_to_all_blocks" ?
If true, we concatenate to all blocks; if false, just to the first one?


How about skip_connection_action = "avg" / "add" ?

[the remark this answers is not in the document]
Comment (mstadelmann) [reply src/chuchichaestli/models/unet/unet.py:287]: (And maybe add a comment somewhere explaining how skip connections are handled with multiple blocks - not that it looks like a bug if someone feels it should behave differently.)

File: src/chuchichaestli/models/unet/time_embeddings.py [hunk 2]
```diff
@@ -39,16 +39,16 @@ def __init__(
         self,
         embedding_size: int = 256,
         scale: float = 1.0,
-        log=True,
-        flip_sin_to_cos=False,
+        log: bool = True,
+        flip_sin_to_cos: bool = False,
```
Comment (mstadelmann) [inline src/chuchichaestli/models/unet/time_embeddings.py:43]: Expose this as a parameter in the Unet class?

**Comment by mstadelmann** [comment]
Otherwise it looks good - all the issues I had fixed on my side are already taken into account now.

**Comment by phdenzel** [comment]
Alright. I'll apply what we discussed above, and once finished, undraft the PR for final review.

**Comment by phdenzel** [comment]
### Summary of changes
- fixed terminology in docstring and constructor argument descriptions
- adjusted keyword arguments to match terminology (removed mentions of layers in favour of blocks to avoid confusion)
  - `num_layers_per_block` -> `num_blocks_per_level`
  - `skip_connection_to_all_layers` -> `skip_connection_to_all_blocks`
- added multiple time embedding choices: `SinusoidalTimeEmbedding` or `DeepSinusoidalTimeEmbedding`
- exposed `t_emb_*` keyword arguments for fine-grained adjustment
- dedup of `range(num_blocks_per_level)`-loop
- `skip_connection_to_all_blocks=True` links the last block in the encoder to all blocks in the corresponding level of the decoder
- filled various gaps in test coverage for `unet.py`, `blocks.py`, and `time_embeddings.py`

**Review by phdenzel — APPROVED** [review]
After a minor bug fix in `UNet.forward`, all seems fine now.

**Review by phdenzel — DISMISSED** [review]

File: src/chuchichaestli/models/resnet.py [hunk 3]
```diff
@@ -74,11 +74,11 @@ def __init__(
 
         self.dropout = nn.Dropout(res_dropout)
 
-    def forward(self, x: torch.Tensor, t: torch.Tensor) -> torch.Tensor:
+    def forward(self, x: torch.Tensor, t: torch.Tensor | None = None) -> torch.Tensor:
         """Forward pass through the residual block."""
         hh = self.conv1(self.act1(self.norm1(x)))
         idx = [slice(None), slice(None)] + [None] * self.dimensions
-        if self.time_embedding:
+        if self.time_embedding and t is not None:
```
Comment (mstadelmann) [inline src/chuchichaestli/models/resnet.py:81]: Is time_embedding=True and t not none a valid case?
-> raise an error instead?

File: src/chuchichaestli/models/unet/blocks.py [hunk 4]
```diff
@@ -115,10 +113,10 @@ def __init__(
         self,
         dimensions: int,
         channels: int,
-        time_embedding: bool = True,
+        time_embedding: bool = False,
```
Comment (mstadelmann) [inline src/chuchichaestli/models/unet/blocks.py:116]: Finally this is fixed as well :-)

File: src/chuchichaestli/models/unet/unet.py [hunk 5]
```diff
@@ -59,115 +68,200 @@
...
+    The decoder is built symmetrically to the encoder with (residual) transposed
+    convolutional and upsampling blocks, each level linked via skip connections
+    which ensure spatial information is passed through the network.
+    """
 
     def __init__(
         self,
         dimensions: int = 2,
         in_channels: int = 1,
         n_channels: int = 32,
```
[the remark this answers is not in the document]
Comment (mstadelmann) [reply src/chuchichaestli/models/unet/unet.py:93]: Or even call it something like `nb_filters_first_block`?
(Having a new user in mind... this might help?)

File: src/chuchichaestli/models/unet/unet.py [hunk 6]
```diff
@@ -59,115 +68,200 @@
...
+            "leakyrelu,0.2",
+            "softplus",
+        ] = "silu",
         in_kernel_size: int = 3,
         out_kernel_size: int = 3,
+        time_embedding: Literal[
+            "SinusoidalTimeEmbedding", "DeepSinusoidalTimeEmbedding"
+        ]
+        | bool
+        | None = None,
```
Comment (mstadelmann) [inline src/chuchichaestli/models/unet/unet.py:137]: This is ruff formatting? Looks weird... :-)

File: src/chuchichaestli/models/unet/unet.py [hunk 7]
```diff
@@ -59,115 +68,200 @@
...
         attn_dropout_p: float = 0.0,
-        attn_norm_type: str = "group",
+        attn_norm_type: Literal["group", "instance", "batch", "adabatch"] = "group",
         attn_groups: int = 32,
         attn_kernel_size: int = 1,
         attn_gate_inter_channels: int = 32,
-        skip_connection_action: str = "concat",
-        skip_connection_between_levels: bool | None = None,
-        add_noise: str | None = None,
+        skip_connection_action: Literal["concat", "avg", "add"] = "concat",
```
Comment (mstadelmann) [inline src/chuchichaestli/models/unet/unet.py:179]: Add None to typing?

File: src/chuchichaestli/models/unet/unet.py [hunk 8]
```diff
@@ -59,115 +68,200 @@
...
+        ] = "silu",
         in_kernel_size: int = 3,
         out_kernel_size: int = 3,
+        time_embedding: Literal[
+            "SinusoidalTimeEmbedding", "DeepSinusoidalTimeEmbedding"
+        ]
+        | bool
+        | None = None,
+        time_channels: int = 32,
+        t_emb_dim: int = 32,
```
Comment (mstadelmann) [inline src/chuchichaestli/models/unet/unet.py:139]: `time_channels` and `t_emb_dim` are risky to confuse. But again, it's documented below, so I guess it's ok.

File: src/chuchichaestli/models/unet/unet.py [hunk 9]
```diff
@@ -59,115 +68,200 @@
...
         ),
-        block_out_channel_mults: tuple[int, ...] = (1, 2, 2, 4),
-        time_embedding: bool = True,
-        time_channels: int = 32,
-        num_layers_per_block: int = 1,
+        block_out_channel_mults: Sequence[int] = (1, 2, 2, 4),
+        upsample_type: Literal["Upsample", "UpsampleInterpolate"] = "Upsample",
+        downsample_type: Literal["Downsample", "DownsampleInterpolate"] = "Downsample",
+        num_blocks_per_level: int = 1,
         groups: int = 8,
```
Comment (mstadelmann) [inline src/chuchichaestli/models/unet/unet.py:118]: This could use a bit more detail.

`groups: Number of groups for group normalization.` -> What settings must be set for this to take action? How does it relate to res_groups and attn_groups? 

And lets move it next to the norm type param?

File: src/chuchichaestli/models/unet/unet.py [hunk 10]
```diff
@@ -59,115 +68,200 @@
...
+        upsample_cls = UPSAMPLE_FUNCTIONS[upsample_type]
+        downsample_cls = DOWNSAMPLE_FUNCTIONS[downsample_type]
+        n_mults = len(block_out_channel_mults)
+        self.num_blocks_per_level = num_blocks_per_level
+        self.skip_connection_to_all_blocks = skip_connection_to_all_blocks
 
+        # Group normalization optimization
         if res_norm_type == "group" and n_channels % res_groups != 0:
             warnings.warn(
                 f"Number of channels ({n_channels}) is not divisible by the number of groups ({res_groups}). Setting number of groups to in_channels."
```
Comment (mstadelmann) [inline src/chuchichaestli/models/unet/unet.py:259]: We could raise an error instead of the warning...?

**Review by mstadelmann — APPROVED** [review]
Looking good - I added a few suggestions. (I did not look at the tests though).

File: src/chuchichaestli/models/unet/unet.py (hunk 10 above)
Comment (phdenzel) [reply src/chuchichaestli/models/unet/unet.py:259]: This  is a decision torch made. The group normalization class automatically clips the groups if the number of channels is smaller than the number of groups. I think warnings are better here, because it just points out that your setting has no effect.

File: src/chuchichaestli/models/resnet.py (hunk 3 above)
Comment (phdenzel) [reply src/chuchichaestli/models/resnet.py:81]: Thinking about it again, I guess not... I'll revert.

**Review by mstadelmann — APPROVED** [review]
